internal/action/actions.go

```diff
@@ -1162,101 +1162,180 @@ func (h *BufPane) Redo() bool {
 	return true
 }
 
+func (h *BufPane) selectLines() int {
+	if h.Cursor.HasSelection() {
+		start := h.Cursor.CurSelection[0]
+		end := h.Cursor.CurSelection[1]
+		if start.GreaterThan(end) {
+			start, end = end, start
+		}
+		if end.X == 0 {
+			end = end.Move(-1, h.Buf)
+		}
+
+		h.Cursor.Deselect(true)
+		h.Cursor.SetSelectionStart(buffer.Loc{0, start.Y})
+		h.Cursor.SetSelectionEnd(buffer.Loc{0, end.Y + 1})
+	} else {
+		h.Cursor.SelectLine()
+	}
+	return h.Cursor.CurSelection[1].Y - h.Cursor.CurSelection[0].Y
+}
+
 // Copy the selection to the system clipboard
 func (h *BufPane) Copy() bool {
-	if h.Cursor.HasSelection() {
-		h.Cursor.CopySelection(clipboard.ClipboardReg)
-		h.freshClip = true
-		InfoBar.Message("Copied selection")
+	if !h.Cursor.HasSelection() {
+		return false
 	}
+	h.Cursor.CopySelection(clipboard.ClipboardReg)
+	h.freshClip = false
+	InfoBar.Message("Copied selection")
 	h.Relocate()
 	return true
 }
 
-// CopyLine copies the current line to the clipboard
+// CopyLine copies the current line to the clipboard. If there is a selection,
+// CopyLine copies all the lines that are (fully or partially) in the selection.
 func (h *BufPane) CopyLine() bool {
-	if h.Cursor.HasSelection() {
+	origLoc := h.Cursor.Loc
+	origLastVisualX := h.Cursor.LastVisualX
+	origSelection := h.Cursor.CurSelection
+
+	nlines := h.selectLines()
+	if nlines == 0 {
 		return false
 	}
-	origLoc := h.Cursor.Loc
-	h.Cursor.SelectLine()
 	h.Cursor.CopySelection(clipboard.ClipboardReg)
-	h.freshClip = true
-	InfoBar.Message("Copied line")
+	h.freshClip = false
+	if nlines > 1 {
+		InfoBar.Message(fmt.Sprintf("Copied %d lines", nlines))
+	} else {
+		InfoBar.Message("Copied line")
+	}
 
-	h.Cursor.Deselect(true)
 	h.Cursor.Loc = origLoc
+	h.Cursor.LastVisualX = origLastVisualX
+	h.Cursor.CurSelection = origSelection
 	h.Relocate()
 	return true
 }
 
-// CutLine cuts the current line to the clipboard
-func (h *BufPane) CutLine() bool {
-	h.Cursor.SelectLine()
+// Cut the selection to the system clipboard
+func (h *BufPane) Cut() bool {
 	if !h.Cursor.HasSelection() {
 		return false
 	}
-	if h.freshClip {
-		if h.Cursor.HasSelection() {
-			if clip, err := clipboard.Read(clipboard.ClipboardReg); err != nil {
-				InfoBar.Error(err)
-			} else {
-				clipboard.WriteMulti(clip+string(h.Cursor.GetSelection()), clipboard.ClipboardReg, h.Cursor.Num, h.Buf.NumCursors())
-			}
+	h.Cursor.CopySelection(clipboard.ClipboardReg)
+	h.Cursor.DeleteSelection()
+	h.Cursor.ResetSelection()
+	h.freshClip = false
+	InfoBar.Message("Cut selection")
+
+	h.Relocate()
+	return true
+}
+
+// CutLine cuts the current line to the clipboard. If there is a selection,
+// CutLine cuts all the lines that are (fully or partially) in the selection.
+func (h *BufPane) CutLine() bool {
+	nlines := h.selectLines()
+	if nlines == 0 {
+		return false
+	}
+	totalLines := nlines
+	if h.freshClip && time.Since(h.lastCutTime) < 10*time.Second {
```

Comment:
Regarding spaces, it is the [gofmt](https://pkg.go.dev/cmd/gofmt) style.



> 2\. I know this is a behavior from before but I never knew it was there.

Neither did I. And I agree with you.

If we are talking about the append-to-clipboard feature as such, I agree we should document and/or make it optional, and I'm inclined to think that we should do both. But it seems better (more flexible) not to add an option but to add a separate action. Rename the current `CutLine` implementation to `CutLineAppend` or `CutLineSmart` (both names suck, but I can't think of anything better at the moment), add a new `CutLine` implementation providing the "classical" line cut functionality without this fancy behavior, and update the default Ctrl-x and Ctrl-k bindings accordingly, to not change their default behavior.

And if we are talking specifically about this `lastCutTime` sub-feature (resetting the append-to-clipboard after 10 minutes), AFAICS it was buggy and never worked from the beginning. (Someone though it was a good idea to implement it, but didn't bother to test it, lol.) And I'm not sure it's a useful feature. So, since it has never really worked anyway, I'm inclined to think we should just remove it.

Reply:
PR that originally added it: #64 + #70.

Reply:
> Regarding spaces, it is the [gofmt](https://pkg.go.dev/cmd/gofmt) style.

It's weird how some places in the same file we have spaces but not this one. I don't code much in golang and it is a very minor thing so I don't mind, just pointing it out.

> I'm inclined to think that we should do both. But it seems better (more flexible) not to add an option but to add a separate action. Rename the current `CutLine` implementation to `CutLineAppend`

Yeah, this seems good to me. `CutLineAppend` is fine I think. 

>So, since it has never really worked anyway, I'm inclined to think we should just remove it.

Yeah, it was never working so we can remove it.

Reply:
> It's weird how some places in the same file we have spaces but not this one.

Apparently there is some logic in that. Like, use spaces in standalone statements but don't use them in function arguments (but only if there is more than one function argument), etc.

I'd prefer to always use spaces, but our rule is: just follow whatever style is expected by `gofmt`, instead of wasting time on discussions which style is better.



> `CutLineAppend` is fine I think.

Or `ClipLine`, maybe?

Reply:
`ClipLine` is good and would exactly reflects what it does.

Anyway, would someone really miss that "feature" in the moment it was never documented, never working and thus never really present to someone? Maybe it is really better to simply clean it up.

Reply:
> never working

It was working. Only the "start afresh after 10 seconds" part wasn't working.

You can try it: press `Ctrl-k` several times and then press `Ctrl-v`.

Reply:
> Now I'm a bit confused that a `Cut` of the whole selected line cuts as expected the content to the clipboard and leaves the line empty

Because you selected the line without selecting the newline character at the end of it?



> Now that I know what it really was about I'd go with something like `CutAndClipLine`, because it's not just copying/appending to the clip, but cutting as well.

Unless we are considering adding a "just copy and append to the clipboard" action as well (I doubt we are), I think `CutAndClipLine` is unnecessarily verbose.

Reply:
I think `ClipLine` doesn't really describe what it is doing. If we have the same thing for copy line (where it appends to the clipboard), the word `ClipLine` is confusing.

I still think `CutLineAppend` (Or some variation of it if you don't like it) describes what it is doing, and we can even have the same naming style for `CopyLine` (that appends to clipboard) as well if we want to

Reply:
Ok, I suggested `ClipLine` because it sounded cool and succinct to me, but frankly I'm not really familiar with the meaning of the word "clip" (I'm not a native speaker). Doesn't it actually have two different meanings: one like "append" but the other one like "cut"?

Anyway... now I'm thinking: maybe let's not complicate things and leave the current `CutLine` behavior as is (just document it _somewhere_; and remove the 10 seconds part). If the user wants the simple behavior instead, the user can just use `CopyLine,DeleteLine`, right?

Reply:
> Doesn't it actually have two different meanings: one like "append" but the other one like "cut"?

It means "cut" to me.  I can see how it can mean "append" but wouldn't click for me unless I used it.

> leave the current `CutLine` behavior as is

Yeah, that's fine with me as well (to be honest I don't use cutline that much :sweat_smile:)

Ideally, the ability to switch off this behavior would be great but I think this behavior works fine for most scenarios. (We can add it if people really want to switch it off)

Reply:
> Because you selected the line without selecting the newline character at the end of it?

Sure it was. It was just my expectation which didn't fit. My fault...

> Anyway... now I'm thinking: maybe let's not complicate things and leave the current `CutLine` behavior as is (just document it somewhere; and remove the 10 seconds part). If the user wants the simple behavior instead, the user can just use `CopyLine,DeleteLine`, right?

Yep, I'm fine with that too, since I didn't used it so far.

Reply:
Removed the `lastCutTime` feature and added documentation for the `CutLine` behavior.